test/Draw.rb

```diff
@@ -0,0 +1,414 @@
+    assert_nothing_raised { @draw.stroke_width = 15 }
+    assert_raise(TypeError) { @draw.stroke_width = 'x' }
+  end
+
+  def test_text_antialias
+    assert_nothing_raised { @draw.text_antialias = true }
+    assert_nothing_raised { @draw.text_antialias = false }
+  end
+
+  def test_tile
+    assert_nothing_raised { @draw.tile = nil }
+    assert_nothing_raised do
+      img1 = Magick::Image.new(10, 10)
+      img2 = Magick::Image.new(20, 20)
+
+      @draw.tile = img1
+      @draw.tile = img2
+    end
+  end
+
+  def test_undercolor
+    assert_nothing_raised { @draw.undercolor = Magick::Pixel.from_color('white') }
+    assert_nothing_raised { @draw.undercolor = 'white' }
+    assert_raise(TypeError) { @draw.undercolor = 2 }
+  end
+
+  def test_annotate
+    assert_nothing_raised do
+      img = Magick::Image.new(10, 10)
+      @draw.annotate(img, 0, 0, 0, 20, 'Hello world')
+
+      yield_obj = nil
+      @draw.annotate(img, 100, 100, 20, 20, 'Hello world 2') do |draw|
+        yield_obj = draw
+      end
+      assert_instance_of(Magick::Draw, yield_obj)
+    end
+
+    assert_raise(TypeError) do
+      img = Magick::Image.new(10, 10)
+      @draw.annotate(img, 0, 0, 0, 20, nil)
+    end
+
+    assert_raise(NoMethodError) { @draw.annotate('x', 0, 0, 0, 20, 'Hello world') }
+  end
+
+  def test_dup
+    @draw.path('M110,100 h-75 a75,75 0 1,0 75,-75 z')
+    @draw.taint
+    @draw.freeze
+    dup = @draw.dup
+    assert_instance_of(Magick::Draw, dup)
+  end
+
+  def test_clone
+    @draw.taint
+    @draw.freeze
+    clone = @draw.clone
+    assert_instance_of(Magick::Draw, clone)
+  end
+
+  def test_composite
+    composite_operators = [
+        Magick::AddCompositeOp,
+        Magick::AtopCompositeOp,
+        Magick::BlendCompositeOp,
+        Magick::BlurCompositeOp,
+        Magick::BumpmapCompositeOp,
+        Magick::ChangeMaskCompositeOp,
+        Magick::ClearCompositeOp,
+        Magick::ColorBurnCompositeOp,
+        Magick::ColorDodgeCompositeOp,
+        Magick::ColorizeCompositeOp,
+        Magick::CopyBlackCompositeOp,
+        Magick::CopyBlueCompositeOp,
+        Magick::CopyCompositeOp,
+        Magick::CopyCyanCompositeOp,
+        Magick::CopyGreenCompositeOp,
+        Magick::CopyMagentaCompositeOp,
+        Magick::CopyOpacityCompositeOp,
+        Magick::CopyRedCompositeOp,
+        Magick::CopyYellowCompositeOp,
+        Magick::DarkenCompositeOp,
+        Magick::DarkenIntensityCompositeOp,
+        Magick::DistortCompositeOp,
+        Magick::DivideCompositeOp,
+        Magick::DivideSrcCompositeOp,
+        Magick::DstAtopCompositeOp,
+        Magick::DstCompositeOp,
+        Magick::DstInCompositeOp,
+        Magick::DstOutCompositeOp,
+        Magick::DstOverCompositeOp,
+        Magick::DifferenceCompositeOp,
+        Magick::DisplaceCompositeOp,
+        Magick::DissolveCompositeOp,
+        Magick::ExclusionCompositeOp,
+        Magick::HardLightCompositeOp,
+        # Magick::HardMixCompositeOp,
```

Comment:
Maybe this can be added conditional by doing the code below in `test_all_basic.rb`:

```ruby
IM_VERSION = Gem::Version.new(Regexp.last_match(1))
IM_REVISION = Gem::Version.new(Regexp.last_match(2))
IM_689 = IM_VERSION >= Gem::Version.new('6.8.9')
```

And then do this here:

````ruby
if IM_689
  composite_operators = composite_operators + [Magick::HardMixCompositeOp]
end
````

Reply:
I think we should provide a way to retrieve runtime ImageMagick version.

Reply:
Thanks for adding this 👍 . I do think that the check should be done with a variable defined inside `test_all_basic.rb` so we can use this across multiple tests/files.